server/src/main/kotlin/fi/oph/kitu/kotoutumiskoulutus/KoealustaService.kt

```diff
@@ -34,7 +35,15 @@ class KoealustaService(
         try {
             return jacksonObjectMapper.enable(JsonParser.Feature.INCLUDE_SOURCE_IN_LOCATION).readValue<T>(json)
         } catch (e: Exception) {
-            val moodleError = jacksonObjectMapper.readValue<MoodleErrorMessage>(json)
+            val moodleError =
```

Comment:
Nopealla lukemisella tämä on vähä hassu et mitä tässä tehdään. Catchataan virhe ja heitetään uudelleen. Yai sit heitetään Moodle -virhe. 

Joko kapsuloisin tän johonkin `fun throwExceptionInMoodle` - metodiin, tai lisäisin kommentin, joka selittää miks näin pitää tehdä.

Reply:
Jep, tässä tehdään oikeastaan lähtökohtaisesti vähän väärää asiaa, minkä takia tämä poikkeuskäsittely näyttää niin oudolta.

Alkuperäinen poikkeus voi johtua kahdesta eri syystä:

1. Parsittava sisältö onkin Moodlen virheviesti eikä `T`, tai
2. Sisältö ei ole virheviesti mutta ei myöskään ole `T`.

Ensimmäinen tapaus voi johtua mistä vain. Jälkimmäinen tilanne on virhe meidän Moodle-viestin tyypityksessä.

Nämä olisi hyvä käsitellä erikseen, mutta nyt yritetään lukea sokkona `T`, ja jos se ei onnistu, meidän pitää erotella tapaukset 1 ja 2. Tapauksessa 1 luku `MoodleError`iksi onnistuu ja heitetään se. Tapauksessa 2 luku `MoodleError`iksi ei onnistu, jolloin ilman tätä hassua `try-catch`-lohkoa heitetään `MissingKotlinParameterException` ja alkuperäinen poikkeus (`T`:ksi lukemisen virhe) katoaa. Tällä hassulla koodilla varmistetaan, että lokiin tulee myös tuo alkuperäinen virhe, jolloin nähdään mitä yritettiin parsia.

Reply:
Tein tälle vähän siivousta, toivottavasti on ymmärrettävämpi.